### PR TITLE
🌱 Change registry from k8s.gcr.io to registry.k8s.io

### LINKED ISSUE #7402: Change registry in published manifests to registry.k8s.io
Hey folks,

I think it is time that we publish manifests with the new registry location. I did a quick test just to double check and our v1.2.3 images are already available under registry.k8s.io/cluster-api.

While we're on it. Let's check the whole repo for occurences of k8s.gcr.io and see if we can replace them all

Let's also add this to the v1.2=>v1.3 migration guide as a hint for providers.

Let's be careful about just search/replacing k8s.gcr.io across godoc / unit tests / ... . Some of them maybe shouldn't change.

/kind cleanup
/area release

### PR DESCRIPTION
**Pull request by oscr**
**What this PR does / why we need it**:

This pr replaces all `k8s.gcr.io` with `registry.k8s.io`. I remember trying this a while back and running into issues due to some images not being available. But now it seems to work for the none malformed image registries.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes  #7402

### CODE REVIEW

**Comment by fabriziopandini** [comment]
This needs some time for review, I will try to make a first pass before KubeCon

File: controlplane/kubeadm/internal/workload_cluster_coredns.go [hunk 1]
```diff
@@ -49,7 +49,7 @@ const (
 
 	// kubernetesImageRepository is the default Kubernetes image repository for build artifacts.
 	kubernetesImageRepository    = "registry.k8s.io"
-	oldKubernetesImageRepository = "k8s.gcr.io"
+	oldKubernetesImageRepository = "registry.k8s.io"
```
Comment (ameukam) [inline controlplane/kubeadm/internal/workload_cluster_coredns.go:52]: I think we want to keep this value to `k8s.gcr.io` since we `kubernetesImageRepository` to represent the new registry.

Comment (oscr) [reply controlplane/kubeadm/internal/workload_cluster_coredns.go]: Thanks @ameukam! I agree that was a mistake. Fixed

File: bootstrap/kubeadm/types/upstreamv1beta1/types.go [hunk 2]
```diff
@@ -107,8 +107,8 @@ type ClusterConfiguration struct {
 	CertificatesDir string `json:"certificatesDir,omitempty"`
 
 	// ImageRepository sets the container registry to pull images from.
```
Comment (sbueringer) [inline bootstrap/kubeadm/types/upstreamv1beta1/types.go:109]: Let's not modify the old kubeadm structs. (this triggers changes in v1alpha3 API types which are basically frozen)

File: controlplane/kubeadm/internal/workload_cluster.go [hunk 3]
```diff
@@ -86,7 +86,7 @@ var (
 	minVerUnversionedKubeletConfig = semver.MustParse("1.24.0")
 
 	// minKubernetesVersionImageRegistryMigration is first kubernetes version where
-	// the default image registry is registry.k8s.io instead of k8s.gcr.io.
+	// the default image registry is registry.k8s.io instead of registry.k8s.io.
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/workload_cluster.go:89]: Please roll back this change

File: controlplane/kubeadm/internal/workload_cluster.go [hunk 4]
```diff
@@ -627,7 +627,7 @@ func yamlToUnstructured(rawYAML []byte) (*unstructured.Unstructured, error) {
 // * "registry.k8s.io" if v1.25 <= version < v1.26 to migrate to the new registry
 // * "" otherwise.
 // Beginning with kubernetes v1.25, the default registry for kubernetes is registry.k8s.io
-// instead of k8s.gcr.io which is why references should get migrated when upgrading to v1.25.
+// instead of registry.k8s.io which is why references should get migrated when upgrading to v1.25.
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/workload_cluster.go:630]: Same here

File: controlplane/kubeadm/internal/workload_cluster_coredns.go [hunk 5]
```diff
@@ -231,7 +231,7 @@ func (w *Workload) getCoreDNSInfo(ctx context.Context, clusterConfig *bootstrapv
 		return nil, err
 	}
 
-	// Handle the renaming of the upstream image from "k8s.gcr.io/coredns" to "k8s.gcr.io/coredns/coredns"
+	// Handle the renaming of the upstream image from "registry.k8s.io/coredns" to "registry.k8s.io/coredns/coredns"
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/workload_cluster_coredns.go:234]:
```suggestion
	// Handle the renaming of the upstream image from:
	// * "registry.k8s.io/coredns" to "registry.k8s.io/coredns/coredns" or
	// * "k8s.gcr.io/coredns" to "k8s.gcr.io/coredns/coredns"
```

File: controlplane/kubeadm/internal/workload_cluster_coredns_test.go [hunk 6]
```diff
@@ -82,7 +82,7 @@ func TestUpdateCoreDNS(t *testing.T) {
 				Spec: corev1.PodSpec{
 					Containers: []corev1.Container{{
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/workload_cluster_coredns_test.go:83]: Please revert all those test cases. The logic shouldn't be changed (and isn't) the same applies to the test cases

File: controlplane/kubeadm/internal/workload_cluster_test.go [hunk 7]
```diff
@@ -124,21 +124,21 @@ func TestUpdateKubeProxyImageInfo(t *testing.T) {
 			name:        "succeeds if patch correctly",
 			ds:          newKubeProxyDS(),
 			expectErr:   false,
```
Comment (sbueringer) [inline controlplane/kubeadm/internal/workload_cluster_test.go:126]: Same here

File: docs/book/src/developer/providers/v1.2-to-v1.3.md [hunk 8]
```diff
@@ -63,3 +63,4 @@ The default value is 0, meaning that the volume can be detached without any time
   The variable `SkipUpgrade` could be set to revert to the old behaviour by making use of the `KUBERNETES_VERSION` variable and skipping the kubernetes upgrade.
 - cert-manager upgraded from v1.9.1 to v1.10.0.
 - Machine `providerID` is now being strictly checked for equality when compared against Kubernetes node `providerID` data. This is the expected criteria for correlating a Cluster API machine to its corresponding Kubernetes node, but historically this comparison was not strict, and instead compared only against the `ID` substring part of the full `providerID` string. Because different providers construct `providerID` strings differently, the `ID` substring is not uniformly defined and implemented across providers, and thus the existing `providerID` equality can not guarantee the correct Machine-Node correlation. It is very unlikely that this new behavior will break existing providers, but FYI: if strict `providerID` equality will degrade expected behaviors, you may need to update your provider implementation prior to adopting Cluster API v1.3.
+- Registry has changed from `k8s.gcr.io` to `registry.k8s.io`.
```
Comment (sbueringer) [inline docs/book/src/developer/providers/v1.2-to-v1.3.md:66]: Let's rephrase and state that we now publish our images to the new registry as the new default registry is registry.k8s.io

(KCP behavior was already changed in other PRs and shouldn't be relevant for providers)

File: test/framework/kubernetesversions/data/debian_injection_script.envsubst.sh.tpl [hunk 9]
```diff
@@ -103,9 +103,9 @@ if [[ "$${KUBERNETES_VERSION}" != "" ]]; then
     systemctl restart kubelet
   fi
   IMAGE_REGISTRY_PREFIX=registry.k8s.io
```
Comment (sbueringer) [inline test/framework/kubernetesversions/data/debian_injection_script.envsubst.sh.tpl:105]: Please roll back changes in this file. We have to keep using the old registry for old Kubernetes versions

File: test/framework/daemonset_helpers.go [hunk 10]
```diff
@@ -42,13 +42,13 @@ func WaitForKubeProxyUpgrade(ctx context.Context, input WaitForKubeProxyUpgradeI
 	parsedVersion, err := semver.ParseTolerant(input.KubernetesVersion)
 	Expect(err).ToNot(HaveOccurred())
```
Comment (sbueringer) [inline test/framework/daemonset_helpers.go:44]: Please revert changes in this file

File: test/framework/deployment_helpers.go [hunk 11]
```diff
@@ -282,8 +282,8 @@ func WaitForDNSUpgrade(ctx context.Context, input WaitForDNSUpgradeInput, interv
 			return false, err
 		}
 
-		// NOTE: coredns image name has changed over time (k8s.gcr.io/coredns,
```
Comment (sbueringer) [inline test/framework/deployment_helpers.go:285]: Please revert

**Review by sbueringer — COMMENTED** [review]
tl;dr there are a few places where we have to keep the old registry.

We will only migrate from the old to the new registry after certain Kubernetes releases, which is why we can't just replace all occurences of the old registry with the new one

Basically the only behavioral change we want is the one in the Makefile (some of the changes like unit tests / doc are fine though)

**Comment by sbueringer** [comment]
Thank you very much!!
(especially for the quick follow-up after the review)

/lgtm
/approve

Trying this out with the beta release today

**Comment by oscr** [comment]
@sbueringer Thank you for your great help and encouragement :pray:

**Comment by sbueringer** [comment]
You're welcome :). Thx for working on this issue :)
